### PR TITLE
[Relay] Partial Eval now Support Interprocedural optimization and has termination check.

### PR DESCRIPTION
**Pull request by MarisaKirisame**
@nhynes @wweic @icemelon9 @junrushao1994 @ZihengJiang  @jroesch @tqchen can you guys review?

### CODE REVIEW

File: include/tvm/relay/expr.h [hunk 1]
```diff
@@ -539,6 +539,25 @@ class TempExprNode : public ExprNode {
 
 RELAY_DEFINE_NODE_REF(TempExpr, TempExprNode, Expr);
 
+class Annotate;
+class AnnotateNode : public ExprNode {
```
Comment (tqchen) [inline include/tvm/relay/expr.h]: Previously we directly use special op as annotation, let us keep that way and keep the expr stable

**Comment by MarisaKirisame** [comment]
@tqchen can you give another round of review?

**Comment by tqchen** [comment]
@MarisaKirisame please rebase. @wweic @jroesch please help review this PR

File: src/relay/pass/partial_eval.cc [hunk 2]
```diff
@@ -702,27 +857,117 @@ class PartialEvaluator : public ExprFunctor<PStatic(const Expr& e, LetList* ll)>
...
+      }
+    };
+    return AnnotateFuncIdMutator(this).VisitExpr(e);
+  }
+
  private:
   Environment env_;
+  Module mod_;
+  std::unordered_map<GlobalVar, PStatic, NodeHash, NodeEqual> gv_map_;
+  /*! Termination checking is done as follow:
```
Comment (wweic) [inline src/relay/pass/partial_eval.cc]: Is there a test that verifies termination checking(no infinite loop)?

File: src/relay/pass/partial_eval.cc [hunk 3]
```diff
@@ -383,15 +388,78 @@ FInterpreter CPUInterpreter() {
...
+Function AsFunc(const Expr& e) {
+  if (e.as<FunctionNode>()) {
+    return Downcast<Function>(e);
+  } else if (const CallNode* c = e.as<CallNode>()) {
+    CHECK(c->op.same_as(WithFuncIdOp()));
+    CHECK_EQ(c->args.size(), 1);
+    return AsFunc(c->args[0]);
+  } else {
+    LOG(FATAL) << "Unknown case";
+    throw;
```
[the remark this answers is not in the document]
Comment (MarisaKirisame) [reply src/relay/pass/partial_eval.cc]: It is there to stop compiler error.

**Comment by MarisaKirisame** [comment]
note to self: dead code elimination dont do inline, so we need an inline pass to make test work again.

File: src/relay/pass/partial_eval.cc (hunk 2 above)
Comment (MarisaKirisame) [reply src/relay/pass/partial_eval.cc]: there is test_map and test_loop. However, none of the test is working right now as the vm merge break this. I will fix them.

File: src/relay/ir/expr.cc [hunk 4]
```diff
@@ -220,8 +220,8 @@ TVM_REGISTER_API("relay._make.Call")
 
 TVM_STATIC_IR_FUNCTOR_REGISTER(IRPrinter, vtable)
 .set_dispatch<CallNode>([](const CallNode* node, tvm::IRPrinter* p) {
-  p->stream << "CallNode(" << node->op << ", " << node->args << ", "
-    << node->attrs << ", " << node->type_args << ")";
+    p->stream << "CallNode(" << node->op << ", " << node->args << ", "
```
[the remark this answers is not in the document]
Comment (MarisaKirisame) [reply src/relay/ir/expr.cc]: It does not work - the result is completely different from all other files.

[the remark this answers is not in the document]
Comment (MarisaKirisame) [reply src/relay/ir/expr.cc]: Probably due to we use different rules for macro.

File: src/relay/pass/partial_eval.cc (hunk 2 above)
Comment (MarisaKirisame) [reply src/relay/pass/partial_eval.cc]: They are fixed.

**Comment by MarisaKirisame** [comment]
@nhynes @icemelon9 can you guys help review?

**Comment by MarisaKirisame** [comment]
@tqchen @jroesch @slyubomirsky can you guys give it a round of review?

File: src/relay/pass/partial_eval.cc [hunk 5]
```diff
@@ -383,15 +388,78 @@ FInterpreter CPUInterpreter() {
   return CreateInterpreter(Module(nullptr), CPUContext(), target);
 }
 
+bool IsAtomic(const Expr& e) {
+  return e.as<VarNode>() || e.as<OpNode>() || e.as<ConstructorNode>() || e.as<GlobalVarNode>();
```
Comment (slyubomirsky) [inline src/relay/pass/partial_eval.cc]: Are constant nodes atomic?

**Comment by slyubomirsky** [comment]
I think it would be good to include some explanation as to the use of the creation time mechanism and why it's important for termination checking, as that was all very difficult for me to understand. More varied test cases (given how complicated the feature is) would be good too.

File: src/relay/pass/partial_eval.cc (hunk 5 above)
Comment (MarisaKirisame) [reply src/relay/pass/partial_eval.cc]: They arent: they are big.

**Comment by MarisaKirisame** [comment]
@slyubomirsky it is not complicated at all! the high level idea is that the argument must be strictly decreasing in some order to be terminating. We choose the order to simply be the order it is created. This work for lots of the case, because for any adt, the outer node is newer then the inner node. So, as long as any of the argument is decreasing, it is good. To include the case where argument a decrease, argument b increase, then argument b increase, argument a decrease, we specify that decrease mean 'globally decrease' (etc: smaller then all seen time at that argument).
I will look at the comment and update it.

**Comment by MarisaKirisame** [comment]
@slyubomirsky I revisited the comment and couldnt figure out how to change it. Anything that is particularly confusing?

**Comment by slyubomirsky** [comment]
I didn't mean that anything is confusing in itself, just that there should somewhere be a note explaining what the creation times are for and what it does. It took me a fair bit of reading and rereading and thinking to figure out how it worked and that is probably not good for future maintainability.

**Comment by wweic** [comment]
@MarisaKirisame How do we prevent infinite loop of PE? One case I can think of is a function with 2 arguments, the series of created_time that can cause infinite loop is `(1, 1), (0, 2),(2, 1), (1, 3), (3, 2), (2, 4), ...`. 1st argument grow its created time by 1 at odd position, 2nd argument grow its created time by 1 at even position, so it can trick the PE to recurse all the time. 
1st argument:   1, 0, 2, 1, 3, 2, 4, 3, 5, 4
2nd argument: 1, 2, 1, 3, 2, 4, 3, 5, 4, 6

Do you think is this case possible and how do we prevent that?

**Comment by MarisaKirisame** [comment]
> the series of created_time that can cause infinite loop is (1, 1), (0, 2),(2, 1), (1, 3), (3, 2), (2, 4), ...
the lowest_bound is (1, 1), (0, 1), then (2, 1) is checked and found that it does not decrease any argument. I specifically design it to avoid this. I had no idea how to fix the comment though. Can you make some suggestion?

**Comment by MarisaKirisame** [comment]
@wweic I had add some test to make sure. They did pass.

**Comment by slyubomirsky** [comment]
Thank you for expanding some of the comments

**Comment by wweic** [comment]
> the lowest_bound is (1, 1), (0, 1), then (2, 1) is checked and found that it does not decrease any argument.

@MarisaKirisame Got it. I missed that you compare timestamp with historical minimum.

File: src/relay/pass/partial_eval.cc [hunk 6]
```diff
@@ -529,10 +635,47 @@ class PartialEvaluator : public ExprFunctor<PStatic(const Expr& e, LetList* ll)>
...
+          FuncId fid = func_map_.at(func);
+          auto recurse = [&]() {
+            TimeFrame tf(this, fid, args_time);
+            return VisitExpr(RegisterFuncId(TypeSubst(AnnotateFuncId(func->body), subst)), ll);
+          };
+          if (time_map_.count(fid) == 0) {
+            return recurse();
+          } else {
+            // We check to see that at least one argument decrease
+            // with respect to all previous invocation.
```
Comment (wweic) [inline src/relay/pass/partial_eval.cc]: Maybe expand the comment here that the number of recursive inline calls is bounded by the first created time of all arguments(product of created times).

File: src/relay/pass/partial_eval.cc [hunk 7]
```diff
@@ -648,8 +648,11 @@ class PartialEvaluator : public ExprFunctor<PStatic(const Expr& e, LetList* ll)>
           if (time_map_.count(fid) == 0) {
             return recurse();
           } else {
-            // We check to see that at least one argument decrease
-            // with respect to all previous invocation.
+            /* We check to see that at least one argument decrease
+             * with respect to all previous invocation.
+             * The depth of the recursion is bounded by
+             * the sum of the time of all argument at the first call.
```
Comment (wweic) [inline src/relay/pass/partial_eval.cc]: I think it's product in stead of sum?

Comment (MarisaKirisame) [reply src/relay/pass/partial_eval.cc]: it is sum. everytime the sum must decrease by at least 1.

Comment (wweic) [reply src/relay/pass/partial_eval.cc]: yes you are right. @MarisaKirisame

**Pull request merged**